### ISSUE TITLE
Support deriving typeclasses with an effectful return type

### ISSUE THREAD

**Issue by Fristi** [propensive/magnolia#43]
Like a `Decoder[A]` returning a `Either[String, A]`, for each member of a type class we need to decode it. Currently `construct` is deficient to do that.

**Current**
```
def construct[Return](makeParam: Param[Typeclass, Type] => Return): Type
```

How should this look like for effects?
```
def construct[Effect[_] : Applicative, Return](makeParam: Param[Typeclass, Type] => Effect[Return]): Type
```

Yet this would imply that for type classes with multiple methods the effect type will be the same (like codecs, in scodec this holds though) 

Where do we get `Applicative` from? I wouldn't favor depending on cats or scalaz

**Comment by propensive** [comment]
I don't think the API needs to be this specific, does it? Could you work with a new `construct` method which takes a `Seq[Any]` of the parameters? i.e.
```scala
def construct[Return](paramValues: Seq[Any]): Type
```
You could then fold the `ctx.parameters` however you like, to get the effect and then pass the `Seq[Any]` of the parameters to the new `construct`.

**Comment by sirthias** [comment]
+1 for adding something like

```scala
def construct(paramValues: Seq[Any]): Type
```

I'm currently trying to write a "conserving tree rewriter" typeclass that is supposed to return the original case class instance if all its members have been rewritten to the identical instance.
With this new `construct` overload this would be easy. The current design makes this quite hard.

**Comment by fommil** [comment]
oh, I just fixed this in #120

**Comment by Fristi** [comment]
I think this is already done right?
